README.md
# Python YouUpload :: README

Python YouUpload is a script that uploads and schedules a video along with a custom thumbnail to YouTube and adds the new video to an existing playlist (optional) using OAuth 2.0 authentication. Made to eliminate repetitive actions (copy-pasting description, tags, etc.) during frequent YouTube uploads. This is primarily an educational project to learn about the YouTube Data API and OAuth procedures, but with utility in mind.

First use will prompt the user to log into the relevant google account in order to give the app credentials. These will then be stored as a local .pickle file which will be used retrieve tokens for future program execution. The script will refresh tokens as needed and update the .pickle file. Depending on permission expiry, renewed manual logins may be needed. 

The script does not currently come with error handling. Moreover, there is currently no possibility in the YouTube Data API to add information about video elements (such as end screens), such that these still need to be added manually after the script upload.

## Requirements

* Register your app with Google. Find the procedure here: https://developers.google.com/youtube/registering_an_application
* The scopes the script uses are "https://www.googleapis.com/auth/youtube.readonly", "https://www.googleapis.com/auth/youtube.upload" and "https://www.googleapis.com/auth/youtube", so make sure those scopes are added this the app's authentication. It may be that .readonly is not necessary at this stage, but if you plan on retrieving additional information, such as various playlist or video IDs, it can be handy to keep .readyonly. 
* In order to use the OAuth 2.0 steps with this script, a client_secrets.json file is required that contains information from the API Console. The file should be in the same directory as the script. Once downloaded from the API console, move the .json file into the local script folder (same place as script.py) and rename it so "client_secret.json" - the script will look for a file with this _exact name_ in the local directory. 


## Installing
* Copy project files into a new directory (using, for example, ```git clone https://github.com/Firgrep/python-you-upload.git```).
* It is _recommended_ to set up a virtual environment for the script and its dependencies. You can do this by making a new directory and then, when within that directory, run ```py -3 -m venv .venv``` and then open your code editor within that directory (or, if you're already in a code editor, open a new terminal to start using the virtual environment). 
    - If for some reason the virtual environment does not activate, manual activation is possible but it may differ depending on your machine and code editor:
        - On Unix or MacOS, using the bash shell: source /path/to/venv/bin/activate
        - On Unix or MacOS, using the csh shell: source /path/to/venv/bin/activate.csh
        - On Unix or MacOS, using the fish shell: source /path/to/venv/bin/activate.fish
        - On Windows using the Command Prompt: path\to\venv\Scripts\activate.bat
        - On Windows using PowerShell: path\to\venv\Scripts\Activate.ps1

* Run this command to install dependencies ```pip install -r requirements.txt```.
    - NOTE: A bug was encountered using the latest packages from the Google Python API client (2023/03/27), specifically to do with the flow object, therefore these exact depency versions were chosen since they function together right out the box. Future releases of the API client will likely patch this (if they haven't already). (For more information about this, see https://github.com/xyou365/AutoRclone/issues/101)


## Usage
This script is used to upload videos to the sPhil channel (https://www.youtube.com/channel/UC8n4PI9Niv8ksUM-NeVx1ig), so you may want to alter the initial data to fit your video upload, but feel free to use it as an example to inspire you!

Setting the parameters:
* Set the correct paths to your payload directory (media files to be uploaded). For this, it is required to set the relative path to the payload directory as the absolute path is set based on where the script is located on your machine. To seperate concerns, the payload folder is initially moved two levels up to move it out of development directories, but you may feel free to put it wherever you like, even within this project directory should you prefer. You may also rename the payload directory to something else or do away with it altogether (untested).
* Set the generic-related data to what you prefer. The idea behind these constants is that they will likely stay unchanged between repeated uploads, eliminating the need to input repeated data. Said otherwise, this is the data you will frequently share between videos.
* Set the file names for the video and thumbnail to be uploaded. The idea is that one will rename the files themselves and dump them into the payload folder and then run the script instead of changing the file names and path in the script. But you may want to change the file names for other reasons so you can do that here. 
* Set the instance-related data--title, scheduled publish data, unique description, etc.--for the particular executition. Two unique description constants are present which will be added to the main description constant, one for the beginning and one for the end respectively. The idea behind this is that although you may have a bulk of data that you will use between video uploads, you may also want a short description particularly for _this_ video and perhaps a credit acknowledgement.
    - Additional note about scheduling video publication: The script is built with scheduling video publications in mind, but if you would like to publish instantly, set ```S_PUBLISH_DATE``` to somewhere in the past (using the same datetime format) and YouTube will publish it instantly upon upload. Do this instead of setting the privacy status to "public". 
* If you would like the new video to be added to a playlist (currently set to be appended as the last element of the list), provide ```S_PLAYLIST_ID``` with a Playlist ID (string). This can be retrieved in the URL information of a playlist after the "=" in https://www.youtube.com/playlist?list={playlistId}. Currently, an ID is retrieved from a local machine using environmental variables. You can do the same or just set the string directly. If not playlist is provided, the script will skip this part of the process.

Running the script:
* To run the script, execute in the terminal ```python script.py``` .
* First-time usage will prompt a link to follow in order to gain permissions. Make sure you are logged into the relevant Google account on the browser and select it from the list. Once permissions have been granted, return to the terminal and copy-paste the required authentication key (try right-clicking with the mouse to paste contents) and press enter. If successful, the script will run through its requests and print to the console when it completes its steps. 
* When the script is done, the video and thumbnail have been uploaded to YouTube. 


## Errors
While no error handling has been programmed, the google api client is usually pretty good printing to the terminal what is wrong in the case of an error. For more info on this, see https://developers.google.com/youtube/v3/docs/errors .


## Resources
This script was build with the use of the resources below. 
* YouTube Data API Documentation. Includes guides, samples and references. https://developers.google.com/youtube/v3 
* OAuth tutorial by Corey Schafer. Extremely helpful in understanding how OAuth work and how to pickle credentials with Python. https://www.youtube.com/watch?v=vQQEaSnQ_bs 


CHANGELOG.md
# Changelog

* 1.1
    - Added resumable upload to the video upload of the script.

* 1.0
    - Working functionality with authentication and upload of video, thumbnail and adding to playlist. 


script.py
# -*- coding: utf-8 -*-
import os
import pickle

import google_auth_oauthlib.flow
import googleapiclient.discovery
import googleapiclient.errors
from googleapiclient.http import MediaFileUpload
from google.auth.transport.requests import Request

# Path-related top-level
ABSOLUTE_PATH = os.path.dirname(__file__)
RELATIVE_PATH = "../../payload"
FULL_PATH = os.path.join(ABSOLUTE_PATH, RELATIVE_PATH)

# Instance-related data - information unique to this video upload.
S_TITLE = "Hegel on the Sciences, Arts and Religion in Ancient China"
S_PUBLISH_DATE = "2023-03-29T19:20+01:00"
S_UNIQUE_DESC_START = "Ahilleas and Filip discuss the fourth and final part concerning the history of China as a moment of Hegel's account of world history. Pertinent here are the sciences, arts and the religion."
S_UNIQUE_DESC_END = ""

# Generic-related data - information shared between videos of which this is an instance.
F_UPLOAD_MEDIA = "vid.mp4" #Easier to leave these as is and alter the name of the files instead.
F_THUMBNAIL = "t.png"
S_PRIVACY_STATUS = "private"
S_CATEGORY_ID = "27"
B_FOR_KIDS = False
S_DESCRIPTION = f"""
{S_UNIQUE_DESC_START}

📘 The text used is GWF Hegel, Lectures on the Philosophy of World Hisory Volume 1: Manuscripts of the Introduction and The Lectures of 1822-3, Edited and Translated by Robert F. Brown and Peter C. Hodgson, with the assistance of William G. Geuss. CLARENDON PRESS • OXFORD, 2011. ISBN 978-0-19-960170-7.

=======================================

AHILLEAS ROKNI completed his PhD thesis in philosophy in 2022, under the supervision of Professor Stephen Houlgate, at the University of Warwick. His thesis aimed to give an account of the much-debated move from the Science of Logic to the Philosophy of Nature in Hegel’s system. Ahilleas’s main research concerns are Hegel’s Logic, philosophy of nature, philosophy of science, and aesthetics. Ahilleas’ overarching interest is in the cohesiveness of Hegel’s systematic philosophy. Hegel’s philosophy is not just the Science of Logic or the Philosophy of Right but is an account of how abstract metaphysics immanently develops into concrete reality This was the impetus behind his PhD research and it is the impetus behind his interest in sPhil: to continue to interrogate whether Hegel’s systematic philosophy is coherent.

🎙️ Ahilleas also hosts the podcast Please Expand that centers on interviews of authors of non-fiction books on topics ranging from history to philosophy, and science to religion, which you can find here: https://www.pleaseexpand.com/

💫 Along with Filip and others, Ahilleas has organized the Hegel at Warwick conference series. https://hegelwarwick.wordpress.com/


FILIP NIKLAS completed his PhD in philosophy in 2022, under the supervision of Professor Stephen Houlgate, at the University of Warwick. The title of his thesis was Hegel’s Critique of Determinism: Justifying Unfreedom as a Moment of Freedom.

👨🏻‍🏫 Filip gives online courses at the Halkyon Academy. You can check out his and other courses in the Cathedral of Learning here: https://www.halkyonguild.org/

📗 For his course on Hegel’s PHENOMENOLOGY OF SPIRIT, please go to: https://halkyonacademy.teachable.com/p/hegel-s-phenomenology-of-spirit
📚 For his course on Hegel’s SYSTEM more broadly, please go to:
https://halkyonacademy.teachable.com/p/hegel-masterclass1
📘 For his course on WHAT IS PHILOSOPHY – an introduction to philosophy through R. G. Collingwood and others – please go to: 
https://halkyonacademy.teachable.com/p/what-is-philosophy

Filip also enjoys ensnaring words in poetic thought-patterns: You can find some of his work here:
📒 https://tinyurl.com/saedah-poem

📲 http://www.filipniklas.com/
🐦@FilipNiklas
=======================================

{S_UNIQUE_DESC_END}
"""

L_TAGS = ["Philosophy","hegel","logic","ontology","metaphysics","german idealism", "idealism", "existence","freedom","categories","Houlgate","thinking","thought","speculative" "thinking","conceptual" "thinking","conceptual" "analysis","concepts","world history", "history"]
S_PLAYLIST_ID = os.environ.get("SPHIL_WORLDHISTORY_PLAYLIST_ID")

# Scopes - part of authentication call.
scopes = ["https://www.googleapis.com/auth/youtube.readonly", "https://www.googleapis.com/auth/youtube.upload", "https://www.googleapis.com/auth/youtube"]

def main():
    credentials = None

    # Disable OAuthlib's HTTPS verification when running locally.
    # *DO NOT* leave this option enabled in production.
    os.environ["OAUTHLIB_INSECURE_TRANSPORT"] = "1"

    api_service_name = "youtube"
    api_version = "v3"
    client_secrets_file = "client_secret.json"

    # AUTHENTICATION
    # First retrieve token.pickle, which stores the user's credentials from previously successful logins.
    if os.path.exists("token.pickle"):
        print("Loading Credentials From File...")
        with open("token.pickle", "rb") as token:
            credentials = pickle.load(token)

    # If no valid credentials availale, refresh token or log in.
    if not credentials or not credentials.valid:
        if credentials and credentials.expired and credentials.refresh_token:
            print("Refreshing Access Token...")
            credentials.refresh(Request())
            with open("token.pickle", "wb") as f:
                print("Saving Refreshed Credentials for Future Use...")
                pickle.dump(credentials, f)

        else: 
            print("Fetching New Tokens...")
            flow = google_auth_oauthlib.flow.InstalledAppFlow.from_client_secrets_file(
                    client_secrets_file, scopes)
            credentials = flow.run_local_server(port=0)
            with open("token.pickle", "wb") as f:
                print("Saving Credentials for Future Use...")
                pickle.dump(credentials, f)

    # YOUTUBE OBJECT BUILD
    youtube = googleapiclient.discovery.build(
        api_service_name, api_version, credentials=credentials)

    # Video upload.
    media = MediaFileUpload(os.path.join(FULL_PATH, F_UPLOAD_MEDIA), resumable=True)
    request = youtube.videos().insert(
        part="snippet,status",
        body={
          "snippet": {
            "categoryId": S_CATEGORY_ID,
            "description": S_DESCRIPTION,
            "title": S_TITLE,
            "tags": L_TAGS,
          },
          "status": {
            "privacyStatus": S_PRIVACY_STATUS,
            "publishAt": S_PUBLISH_DATE,
            "selfDeclaredMadeForKids": B_FOR_KIDS
          }
        },
        media_body=media
    )
    response = None
    while response is None:
      status, response = request.next_chunk()
      if status:
        print(f"Uploaded {int(status.progress() * 100)}%.")

    videoId = response["id"]
    print("Video Upload Complete.")

    # Thumbnail upload.
    request_second = youtube.thumbnails().set(
        videoId=videoId,
        media_body=MediaFileUpload(os.path.join(FULL_PATH, F_THUMBNAIL))
    )
    request_second.execute()
    print("Thumbnail Upload Complete.")

    # If playlist detected, add to playlist.
    if S_PLAYLIST_ID:
        print("Playlist detected. Adding video to playlist...")
        request_third = youtube.playlistItems().insert(
            part="snippet",
            body={
              "snippet": {
                "playlistId": S_PLAYLIST_ID,
                "resourceId": {
                  "kind": "youtube#video",
                  "videoId": videoId
                }
              }
            }
        )

        request_third.execute()
        print("Video Added To Playlist Complete.")
    
    print(f"Script Done. {S_TITLE} has been uploaded to YouTube.")

if __name__ == "__main__":
    main()